### ISSUE TITLE
Proposition: add getOriginalData method to PlotDataItem

### ISSUE THREAD

**Issue by sem-geologist** [pyqtgraph/pyqtgraph#2065]
### Short description

I tried to move my (inheriting `PlotCurveItem`) spectral objects into inheriting `PlotDataItem`, so that I  could start benefiting from log scale switching on/off. 
My software use Infinite line, and gets interception with spectral object hitherto using `getData` method (of `PlotCurveItem`), however I see that `getData` in `PlotDataItem` gives data after transformations and log operations, and it gives wrong numbers, so my intersection returns wrong numbers.
While it is tempting to use `_dataset.x` and '_dataset.y' of `PlotDataItem`, those are private attributes. It would be nice if there would be a public method with stable API for its retrieval.
I.e. `getOriginalData()`

**Comment by NilsNemitz** [comment]
Hi @sem-geologist ,

That certainly seems possible.
But I am not quite sure if it is a good idea to promise that we will always keep an unchanged copy of the data around in PlotDataItem. Once we declare that part of the API, it would be very hard to undo.

@j9ac9k , @pijyoi how do you see that? I currently cannot think of a way for PlotDataItem to do its job *without* keeping the original data accessible. So I think providing a stable API to retrieve the raw dataset might be reasonable. For PlotCurveItem we probably wouldn't want to make any promises like that.

@sem-geologist , do you have a good argument for us to consider why it is better to pull the data back out of the plot, rather than keeping a reference to the original data around?

**Comment by j9ac9k** [comment]
I'm a little reluctant here, if you look at `PlotDataItem.setData()` it takes a _wide_ variety of data structures, and I'm concerned that `getOriginalData` will obligate us to return the data structure that was passed in, not the data structure pyqtgraph preserves a reference to.

The other oddity here is do we return a deep or shallow copy; if it's a shallow copy and an end user manipulates that array, I'm a bit concerned that we might be on the hook for some unexplained behavior a user can run in that way.  If it's a deep copy we return, there could be memory/performance consequences too.

**Comment by sem-geologist** [comment]
What about different name: `getInitialData`, `getUntransformedData`.

TLDR; what I am trying to achieve:
What I want is a possibility to model the background (two yellow lines for linear and "exp" type) under peak (fat infiniteline with up-arrow) and background positions (InfiniteLines with downward arrows), and change that background while dragging those positions (so one of requirements is being very responsive, and thus why I am using `pyqtgraph`). It works nicely with my previously used `PlotCurveItem` or `PlotDataItem` until not switcing to log scale for Y:
![HussariX-small-bkgd](https://user-images.githubusercontent.com/12324068/140918309-4fed2495-9a42-4d29-81ab-a75cda978b93.gif)

I was using `getData` of SpectrumCurve, but that gets wrong InfiniteLine and SpectrumCurve (based on PlotDataItem) intersection when on log scale, which then makes the background model curve (also PlotDataItem) to be wrongly generated.
The whole workflow (including getting position of spectrum, getting averaged curve of small section, making background)  with spectra is very GUI based/oriented and makes no sense to be implemented in not-GUI parts of code.

If there will be no public function I think will find a work around. I.e. I could register most recent reference to the y and x data of spectra during `setData` called by my own method of `setSpectrumData`  in SpectrumCurve type. x and y can be raw, or different takes and normalized units (i.e. x as wavelength, as sin(theta) as in example, as x-ray energy; y as raw counts per second, or counts, or counts per second per nA). Getting data from Item instead from original not-GUI part "cuts some corners" and produce less delay allowing dragging of InfiniteLine and remaking and repainting the modeled background curve fast enough to not bother or annoy.

**Comment by sem-geologist** [comment]
one more straightforward candidate for method name: `getValues`

**Comment by pijyoi** [comment]
Would it not be possible for you to do something like the following since you are already inheriting from ```PlotDataItem```?
```python
class MyCustomItem(pg.PlotDataItem):
    def setData(self, x, y, *args, **kwargs):
        self._my_raw_data = x, y
        super().setData(x, y, *args, **kwargs)

    def getValues(self):
        return _my_raw_data
```

**Comment by sem-geologist** [comment]
Yes, I can. Actually I am doing currently it even more simple (cutting corners, yeah I know premature optimization is evil...) after calling setData, I add x and y as `.x_ref` and `.y_ref` attributes, as latter I need to address those anyway separately. So I don't need then no separate getter method calls (1)  and no tuple unpacking (2) which would be done many and many times.

My issue is that I would expect possibility of getting raw values by default from PlotDataItem. But I understand this is not so straight forward as there is different kind of data which can be passed to construct `PlotDataItem`, and it is not clear what should be returned if initial data is not the numpy array. I think I am closing this, as if there are multiple doubts and concerns it is rather better to do this in subclass depending on particular application requirements.

**Comment by j9ac9k** [comment]
Hi @sem-geologist thanks for sharing the gif, if your project is publicly accessible in some  fashion, we would love to add it to our [README's Used By section](https://github.com/pyqtgraph/pyqtgraph#used-by).   That looks like quite the application, I would  be eager to hear about how pyqtgraph has held up performance wise for your use case, and if there are any voids in the library (plot/chart types) that are missing that would be beneficial.

> I think I am closing this, as if there are multiple doubts and concerns it is rather better to do this in subclass depending on particular application requirements.

I do think the proposal is good idea, and I do think pyqtgraph should make that data accessible in some fashion; but as I said earlier, I'm usually not great at considering different edge cases, and given that subclassing is the Qt way, I'm inclined to think that if a user needs this functionality, that's the safest way to get it.

**Comment by sem-geologist** [comment]
@j9ac9k , thanks for kind words about that gif. I had called it HussariX https://github.com/sem-geologist/HussariX (Conjugation of Hussaria as a speedy and well organized famous Polish-Lithuanian elite cavalry unit, and X stands for X-rays). The aim is to make a tool which does not lag  and speed is one of the main priorities. If User clicks and interacts with software - it just needs to do what it needs to do in a blink of an eye (unless it is bounded by I/O operations).

I actually finalized the migration from `PlotCurveItem` to `PlotDataItem`, and I see no conceivable difference in speed, and got possibility to set Y into log scale for free. Yeah! I thought I don't need that (I would want sqrt kind of scalling), but now I am pretty convinced that Y-log brings some new insight, something I was not aware before.

Thanks to You I just realized it was missing a proper license file (It is GPL) in the main dir. It is still far from what it is supposed to do, and it is WIP. However, we use it already in a lab near every day as it can improve sensitivity and confidence of measurements (by allowing to come up with a well informed measurement strategy, rather than doing a wild guess) and generally increase workflow of electron probe x-ray microanalysis.

**Comment by j9ac9k** [comment]
> I actually finalized the migration from `PlotCurveItem` to `PlotDataItem`, and I see no conceivable difference in speed, and got possibility to set Y into log scale for free. Yeah! I thought I don't need that (I would want sqrt kind of scalling), but now I am pretty convinced that Y-log brings some new insight, something I was not aware before.

Glad to hear it.  On that note, I believe @NilsNemitz 's work with scaling has set a decent foundation for supporting other non-linear scaling (such as sqrt) mechanisms... I do want to get more of the graphics items to support log-transformations, I believe it was you that brought attention to this with respect to infinite line... 

> Thanks to You I just realized it was missing a proper license file (It is GPL) in the main dir. It is still far from what it is supposed to do, and it is WIP. However, we use it already in a lab near every day as it can improve sensitivity and confidence of measurements (by allowing to come up with a well informed measurement strategy, rather than doing a wild guess) and generally increase workflow of electron probe x-ray microanalysis.

Is any software that's in use not WIP? ;)   Would love a PR linking the project to our README 👍🏻

**Comment by agausmann** [comment]
I believe that this is implemented by #2297 ?
